### PR TITLE
support polling authentication

### PR DESCRIPTION
**Pull request by itboyljm**
Changes:

1. when polling server configure route_table, no need to set port and ip of polling client

2. support polling authentication

3. makes polling authentication configurable

### CODE REVIEW

File: jvm/core/main/scala/com/webank/eggroll/core/util/AuthenticationUtils.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      var encryptText = ""
+      for (i <- 0 until encryptParams.length) {
+        if (i != 0) encryptText += "\n"
+        encryptText += encryptParams(i)
+      }
+      encryptText = new String(encryptText.getBytes, AuthenticationUtils.UTF8)
+      return Base64.getEncoder.encodeToString(AuthenticationUtils.HmacSHA1Encrypt(encryptText, appSecret))
+    } catch {
+      case e: Exception =>
+        e.printStackTrace()
```
Comment (drougon) [inline jvm/core/main/scala/com/webank/eggroll/core/util/AuthenticationUtils.scala:33]: TODO?

File: jvm/roll_site/main/scala/com/webank/eggroll/rollsite/Polling.scala [hunk 2]
```diff
@@ -284,7 +366,49 @@ class DispatchPollingReqSO(eggSiteServicerPollingRespSO: ServerCallStreamObserve
...
+      val heads = new util.HashMap[String, String]()
+      heads.put("TIMESTAMP", timestamp)
+      heads.put("PARTY_ID", authPartyID)
+      heads.put("NONCE", nonce)
+      heads.put("ROLE", role)
+      heads.put("APP_KEY", appKey)
+      heads.put("URI", httpUri)
+      heads.put("SIGNATURE", signature)
+      val body = ""
+      logTrace(s"auth heads:${heads.toString}")
```
Comment (drougon) [inline jvm/roll_site/main/scala/com/webank/eggroll/rollsite/Polling.scala:395]: add rollsite header info to log ?

File: jvm/roll_site/main/scala/com/webank/eggroll/rollsite/Polling.scala [hunk 3]
```diff
@@ -284,7 +366,49 @@ class DispatchPollingReqSO(eggSiteServicerPollingRespSO: ServerCallStreamObserve
...
+
+      val authInterface = RollSiteConfKeys.EGGROLL_ROLLSITE_POLLING_AUTHENTICATOR_INTERFACE.get().toString
+      val splitted = authInterface.split("#")
+      val authenticator = Class.forName(splitted(0)).newInstance()
+      val result = MethodUtils.invokeExactMethod(authenticator, splitted(1), authUrl, heads, body).asInstanceOf[Boolean]
```
Comment (drougon) [inline jvm/roll_site/main/scala/com/webank/eggroll/rollsite/Polling.scala:400]: handle exception and add errcode to log

**Review by drougon — DISMISSED** [review]

File: jvm/core/main/scala/com/webank/eggroll/core/constant/ConfKeys.scala [hunk 4]
```diff
@@ -183,6 +183,15 @@ object RollSiteConfKeys {
   val EGGROLL_ROLLSITE_POLLING_CLIENT_ENABLED = ErConfKey("eggroll.rollsite.polling.client.enabled", false)
   val EGGROLL_ROLLSITE_POLLING_CONCURRENCY = ErConfKey("eggroll.rollsite.polling.concurrency", 50)
   val EGGROLL_ROLLSITE_ROUTE_TABLE_KEY = ErConfKey("eggroll.rollsite.route.table.key", "123")
+  val EGGROLL_ROLLSITE_POLLING_AHTHENTICATION_ENABLE = ErConfKey("eggroll.rollsite.polling.authentication.enable", false)
```
Comment (drougon) [inline jvm/core/main/scala/com/webank/eggroll/core/constant/ConfKeys.scala:186]: EGGROLL_ROLLSITE_POLLING_AHTHENTICATION_ENABLED

**Review by maxwong — APPROVED** [review]

**Review by maxwong — APPROVED** [review]

**Pull request merged**